### PR TITLE
`Development`: Remove deprecations in server code

### PR DESCRIPTION
**Pull request by krusche**
This PR migrates deprecations in server code to the now preferred use, in particular used of `notFound()` and `forbidden()` are changed to the corresponding exceptions.

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

#### Code Review
- [x] Review 1
- [x] Review 2

### CODE REVIEW

**Review by coaxial1 — DISMISSED** [review]
Did look through the changes and didn't find any apparent issues. LGTM 👍

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java [hunk 1]
```diff
@@ -238,35 +217,19 @@ private void validateNewOrUpdatedFileUploadExercise(FileUploadExercise fileUploa
...
-        Optional<FileUploadExercise> optionalFileUploadExercise = fileUploadExerciseRepository.findWithEagerTeamAssignmentConfigAndCategoriesById(exerciseId);
-
-        if (optionalFileUploadExercise.isEmpty()) {
-            return notFound();
-        }
-        FileUploadExercise fileUploadExercise = optionalFileUploadExercise.get();
-
+        var exercise = fileUploadExerciseRepository.findWithEagerTeamAssignmentConfigAndCategoriesById(exerciseId)
+                .orElseThrow(() -> new EntityNotFoundException("FileUploadExercise", exerciseId));
         // If the exercise belongs to an exam, only instructors and admins are allowed to access it, otherwise also TA have access
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:222]:
```suggestion
        // If the exercise belongs to an exam, only editors or above are allowed to access it, otherwise also TA have access
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java [hunk 2]
```diff
@@ -159,7 +169,7 @@ public ResultResource(ProgrammingExerciseParticipationService programmingExercis
         var participation = getParticipationWithResults(planKey);
         if (participation == null) {
             log.warn("Participation is missing for notifyResultNew (PlanKey: {}).", planKey);
-            return notFound();
+            throw new EntityNotFoundException("Participation for build plan" + planKey + " does not exist");
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java:172]:
```suggestion
            throw new EntityNotFoundException("Participation for build plan " + planKey + " does not exist");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java [hunk 3]
```diff
@@ -402,14 +417,7 @@ else if (participation instanceof ProgrammingExerciseParticipation) {
...
         log.debug("REST request to delete Result : {}", resultId);
-        Result result = resultRepository.findByIdElseThrow(resultId);
-        Participation participation = result.getParticipation();
-        if (!participation.getId().equals(participationId)) {
-            return badRequest("participationId", "400",
-                    "participationId of the path doesnt match the participationId of the participation corresponding to the result " + resultId + " !");
-        }
-        Course course = participation.getExercise().getCourseViaExerciseGroupOrCourseMember();
-        authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.TEACHING_ASSISTANT, course, null);
+        Result result = getResultForParticipationAndCheckAccess(participationId, resultId, Role.TEACHING_ASSISTANT);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java:420]: Coday warning: Result is unused. Is this on purpose to only check the access? If yes I think you should just call the method.

Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java:420]: Coday warning: Result is unused. Is this on purpose to only check the access? If yes I think you should just call the method.

File: src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java [hunk 4]
```diff
@@ -227,7 +230,7 @@ void validateExercise(Exercise exercise) throws BadRequestAlertException {
         Result result = resultRepository.findByIdWithEagerFeedbacksElseThrow(resultId);
         Participation participation = submission.getParticipation();
         if (!participation.getId().equals(participationId)) {
-            return badRequest("participationId", "400", "participationId in path does not match the id of the participation to submission " + submissionId + " !");
+            throw new BadRequestAlertException("participationId in path does not match the id of the participation to submission " + submissionId + " !", "Participation", "400");
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java:233]:
```suggestion
            throw new BadRequestAlertException("participationId in path does not match the id of the participation to submission " + submissionId + "!", "Participation", "400");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java [hunk 5]
```diff
@@ -319,15 +329,20 @@ else if (planKey.endsWith("-" + BuildPlanType.SOLUTION.getName())) {
...
+        return new ResponseEntity<>(result, HttpStatus.OK);
+    }
+
+    private Result getResultForParticipationAndCheckAccess(Long participationId, Long resultId, Role role) {
         Result result = resultRepository.findByIdElseThrow(resultId);
         Participation participation = result.getParticipation();
         if (!participation.getId().equals(participationId)) {
-            return badRequest("participationId", "400",
-                    "participationId of the path doesnt match the participationId of the participation corresponding to the result " + resultId + " !");
+            throw new BadRequestAlertException("participationId of the path doesnt match the participationId of the participation corresponding to the result " + resultId + " !",
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java:340]:
```suggestion
            throw new BadRequestAlertException("participationId of the path doesnt match the participationId of the participation corresponding to the result " + resultId + "!",
```

**Review by Strohgelaender — APPROVED** [review]
Overall changes look good, I only added a few smaller comments:

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 6]
```diff
@@ -523,15 +525,15 @@ private boolean validateTemporaryAccessTokenForCourse(String temporaryAccessToke
...
-                ContentDisposition contentDisposition = ContentDisposition.builder("attachment").filename(filename).build();
-                headers.setContentDisposition(contentDisposition);
+                contentDisposition = ContentDisposition.builder("attachment").filename(filename).build();
             }
             else {
-                ContentDisposition contentDisposition = ContentDisposition.builder("inline").filename(filename).build();
-                headers.setContentDisposition(contentDisposition);
+                contentDisposition = ContentDisposition.builder("inline").filename(filename).build();
             }
+            headers.setContentDisposition(contentDisposition);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:536]:
```
ContentDisposition.Builder contentDispositionBuilder;
if (filename.endsWith("htm") || filename.endsWith("html") || filename.endsWith("svg") || filename.endsWith("svgz")) {
    // attachment will force the user to download the file
    contentDispositionBuilder = ContentDisposition.builder("attachment");
} else {
    contentDispositionBuilder = ContentDisposition.builder("inline");
}
headers.setContentDisposition(contentDispositionBuilder.filename(filename).build());
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/lecture/VideoUnitResource.java [hunk 7]
```diff
@@ -92,7 +92,7 @@ public VideoUnitResource(LectureRepository lectureRepository, AuthorizationCheck
             new URL(videoUnit.getSource());
         }
         catch (MalformedURLException exception) {
-            badRequest();
+            return badRequest();
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/lecture/VideoUnitResource.java:95]: Can we directly replace these calls?

File: src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java [hunk 8]
```diff
@@ -79,14 +80,14 @@ public NotificationSettingsResource(NotificationSettingRepository notificationSe
     @PreAuthorize("hasRole('USER')")
     public ResponseEntity<NotificationSetting[]> saveNotificationSettingsForCurrentUser(@NotNull @RequestBody NotificationSetting[] notificationSettings) {
         if (notificationSettings.length == 0) {
-            return badRequest("notificationSettings", "400", "Can not save non existing Notification Settings");
+            throw new BadRequestAlertException("Can not save non existing Notification Settings", "NotificationSettings", "notificationSettingsEmpty");
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:83]:
```suggestion
            throw new BadRequestAlertException("Cannot save non existing Notification Settings", "NotificationSettings", "notificationSettingsEmpty");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ApollonDiagramResource.java [hunk 9]
```diff
@@ -1,6 +1,6 @@
 package de.tum.in.www1.artemis.web.rest;
 
-import static de.tum.in.www1.artemis.web.rest.util.ResponseUtil.conflict;
+import static de.tum.in.www1.artemis.web.rest.util.ResponseUtil.*;
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ApollonDiagramResource.java:3]: Are these changes on purpose? I thought we wanted to avoid star imports. (Note: This happened on multiple files)

File: src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java [hunk 10]
```diff
@@ -170,12 +170,15 @@ public AssessmentResource(AuthorizationCheckService authCheckService, UserReposi
...
+        var isTutor = authCheckService.isAtLeastTeachingAssistantForExercise(exercise, user);
         // It is allowed to get the example assessment, if the user is an instructor or
         // if the user is a tutor and the submission is not used for tutorial in the assessment dashboard
-        boolean isAllowed = authCheckService.isAtLeastInstructorForExercise(exercise)
-                || authCheckService.isAtLeastTeachingAssistantForExercise(exercise) && !Boolean.TRUE.equals(exampleSubmission.isUsedForTutorial());
-        if (!isAllowed) {
-            forbidden();
+        // The reason is that example submissions with isTutorial = false should be shown immediately (with the assessment) to the tutor and
+        // for example submission with isTutorial = true, the assessment should not be shown to the tutor. Instead, the tutor should try to assess it him/herself
+        if (!(isInstructor || isTutor && !exampleSubmission.isUsedForTutorial())) {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java:180]:
```suggestion
        if (!(isInstructor || (isTutor && !exampleSubmission.isUsedForTutorial()))) {
```
This makes it easier to understand to everyone who is not familiar with the order of evaluation

Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java:180]: What about editors?

File: src/test/java/de/tum/in/www1/artemis/VideoUnitIntegrationTest.java [hunk 11]
```diff
@@ -72,13 +72,15 @@ public void testAll_asStudent() throws Exception {
     @Test
     @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
     public void createVideoUnit_asInstructor_shouldCreateVideoUnit() throws Exception {
+        videoUnit.setSource("https://www.youtube.com/embed/8iU8LPEa4o0");
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/VideoUnitIntegrationTest.java:75]: A bit disappointed that you didn't select https://www.youtube.com/embed/dQw4w9WgXcQ ;)

File: src/main/java/de/tum/in/www1/artemis/web/rest/lecture/VideoUnitResource.java (hunk 7 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/lecture/VideoUnitResource.java:95]: sure, you can, but at some point, I wanted to stop and prevent making this PR too big.
In fact, this was a bug before, because the previous code basically did nothing :-(

File: src/main/java/de/tum/in/www1/artemis/web/rest/ApollonDiagramResource.java (hunk 9 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ApollonDiagramResource.java:3]: I wanted to decrease the number of warnings for now, importing deprecated static methods counts as a warning, but does not really help fixing those or having an overview how many issues are left

File: src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java (hunk 10 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java:180]: I don't know, I did not think about the logic too much, just wanted to describe the reason behind it :-)

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java:180]: I changed the variable names to isAtLeastTutor and isAtLeastInstructor. It's more obvious that editors are included in the tutor case then

**Review by Strohgelaender — APPROVED** [review]

**Review by julian-christl — APPROVED** [review]
Changes look good!
